boring/src/ssl/test/mod.rs

```diff
@@ -480,6 +480,76 @@ fn test_alpn_server_unilateral() {
     assert_eq!(None, s.ssl().selected_alpn_protocol());
 }
 
+#[test]
+fn test_select_cert_ok() {
+    let mut server = Server::builder();
+    server
+        .ctx()
+        .set_select_certificate_callback(|_client_hello| Ok(()));
+    let server = server.build();
+
+    let client = server.client();
+    client.connect();
+}
+
+#[test]
+fn test_select_cert_error() {
+    let mut server = Server::builder();
+    server.should_error();
+    server
+        .ctx()
+        .set_select_certificate_callback(|_client_hello| Err(ssl::SelectCertError::ERROR));
+    let server = server.build();
+
+    let client = server.client();
+    client.connect_err();
+}
+
+#[test]
+fn test_select_cert_unknown_extension() {
+    let mut server = Server::builder();
+    let unknown_extension = std::sync::Arc::new(std::sync::Mutex::new(None));
+    server.ctx().set_select_certificate_callback({
+        let unknown = unknown_extension.clone();
+        move |client_hello| {
+            *unknown.lock().unwrap() = client_hello.get_extension(1337).map(ToOwned::to_owned);
+            Ok(())
+        }
+    });
+    let server = server.build();
+
+    let client = server.client();
+    client.connect();
+    assert_eq!(unknown_extension.lock().unwrap().as_deref(), None);
+}
+
+#[test]
+fn test_select_cert_alpn_extension() {
+    let mut server = Server::builder();
+    let alpn_extension = std::sync::Arc::new(std::sync::Mutex::new(None));
+    server.ctx().set_select_certificate_callback({
+        let alpn = alpn_extension.clone();
+        move |client_hello| {
+            *alpn.lock().unwrap() = Some(
+                client_hello
+                    .get_extension(ffi::TLSEXT_TYPE_application_layer_protocol_negotiation as u16)
```

Comment:
Let's expose extension type enum and use it instead of raw ints

Reply:
I don't think there is an extension type enum, the cast is here because the FFI constant is i32.

Reply:
I added a custom fake-enum struct type for the extension type.

Reply:
Yeah, this is what I've meant - it's done in a few other places where integer variant enum in introduced in Rust API for the convenience